rllib/connectors/agent/mean_std_filter.py

```diff
@@ -0,0 +1,136 @@
+
+        filter_shape = tree.map_structure(
+            lambda s: (
+                None
+                if isinstance(s, (Discrete, MultiDiscrete))  # noqa
+                else np.array(s.shape)
+            ),
+            get_base_struct_from_space(ctx.observation_space),
+        )
+        self.filter = MeanStdFilter(filter_shape, demean=True, destd=True, clip=10.0)
+
+    def transform(self, ac_data: AgentConnectorDataType) -> AgentConnectorDataType:
+        d = ac_data.data
+        assert (
+            type(d) == dict
+        ), "Single agent data must be of type Dict[str, TensorStructType]"
+        if SampleBatch.OBS in d:
+            d[SampleBatch.OBS] = self.filter(
+                d[SampleBatch.OBS], update=self._is_training
+            )
+        if SampleBatch.NEXT_OBS in d:
+            d[SampleBatch.NEXT_OBS] = self.filter(
+                d[SampleBatch.NEXT_OBS], update=self._is_training
+            )
+
+        return ac_data
+
+    def to_state_dict(self):
+        return MeanStdObservationFilterAgentConnector.__name__, {
+            "filter": self.filter,
+        }
+
+    @staticmethod
+    def from_state_dict(ctx: ConnectorContext, params: List[Any]):
+        connector = MeanStdObservationFilterAgentConnector(ctx)
+        connector.filter = params["filter"]
+        assert all(ctx.observation_space.shape == connector.filter.shape)
+        return connector
+
+    def reset_state(self) -> None:
+        """Creates copy of current state and resets accumulated state"""
+        if not self._is_training:
+            raise ValueError(
+                "State of {} can only be changed when trainin.".format(self.__name__)
+            )
+        self.filter.reset_buffer()
+
+    def apply_changes(self, other: "Filter", *args, **kwargs) -> None:
+        """Updates self with state from other filter."""
+        # inline this as soon as we deprecate ordinary filter with non-connector
+        # env_runner
+        if not self._is_training:
+            raise ValueError(
+                "Changes can only be applied to {} when trainin.".format(self.__name__)
+            )
+        return self.filter.apply_changes(other, *args, **kwargs)
+
+    def copy(self) -> "Filter":
+        """Creates a new object with same state as self.
+
+        This is a legacy Filter method that we need to keep around for now
+
+        Returns:
+            A copy of self.
+        """
+        # inline this as soon as we deprecate ordinary filter with non-connector
+        # env_runner
+        return self.filter.copy()
+
+    def sync(self, other: "AgentConnector") -> None:
+        """Copies all state from other filter to self."""
+        # inline this as soon as we deprecate ordinary filter with non-connector
+        # env_runner
+        if not self._is_training:
+            raise ValueError(
+                "{} can only be synced when trainin.".format(self.__name__)
+            )
+        return self.filter.sync(other)
```

Comment:
other is a connector, is it better to do ```self.filter.sync(other.filter)``` here, so filter doesn't need to be aware of connector?

Reply:
Yes. This method is so far not used, otherwise, this would have thrown an error. Sorry. Just realized that this is not developer API anymore and I should probably keep unused/untested code completely out of it.

Reply:
oh ok, so the connectors are not synced at this point?

Reply:
They are synchronized, but through [the old mechanism](https://github.com/ray-project/ray/pull/27864/files#diff-380f29c4af45d6f1d4f2b6b51877f4877bb10cbad3e8ddd9cbc048252389dab7R656).
I think as long as we have to have both, filters in their old place and in connectors, it's best to keep the same mechanism.